### ISSUE TITLE
Releasing OSNet code and models

### ISSUE THREAD

**Issue by KaiyangZhou** [KaiyangZhou/deep-person-reid#200]

**Comment by KaiyangZhou** [comment]
Code: https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/osnet.py.
Models: https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html.
Paper: https://arxiv.org/abs/1905.00953.

** Please pull the latest commit. **

How to train OSNet:

```sh
python scripts/main.py \
--root $DATA \
-j 4 \
--app image \
--loss softmax \
--label-smooth \
-s market1501 \
-a osnet_x1_0 \
--height 256 \
--width 128 \
--optim amsgrad \
--lr 0.0015 \
--max-epoch 150 \
--stepsize 60 \
--lr-scheduler single_step \
--batch-size 64 \
--fixbase-epoch 10 \
--open-layers classifier \
--transforms random_flip \
--load-weights saved-models/osnet_x1_0_imagenet.pth \
--save-dir log/osnet_x1_0-market1501-softmax \
--gpu-devices 0
```

Small models such as `osnet_x0_5` and `osnet_x0_25` should be trained longer for better convergence (use `--max-epoch 180 --stepsize 80`).

Same-domain reid performance can be boosted by adding random erasing, e.g. `--transforms random_flip random_erase`. However, be aware that random erasing will substantially decrease the cross-domain performance. This paper shows the first step towards this problem in a more generic setting https://arxiv.org/abs/1805.08974.

**Comment by crossknight** [comment]
Hi @KaiyangZhou,

I trained OSNet with parameters you mentioned above and got these results. 

```
mAP: 33.1%
CMC curve
Rank-1  : 59.6%
Rank-5  : 75.2%
Rank-10 : 80.9%
Rank-20 : 84.9%
```

Any suggestion?

**Comment by KaiyangZhou** [comment]
@crossknight 
what is the full command line you used? did you load the imagenet pretrained model?

**Comment by crossknight** [comment]
@KaiyangZhou 
This is the full command line I used
```
python scripts/main.py \
--root reid-data \
-j 4 \
--app image \
--loss softmax \
--label-smooth \
-s market1501 \
-a osnet_x1_0 \
--height 256 \
--width 128 \
--optim amsgrad \
--lr 0.0015 \
--max-epoch 300 \
--stepsize 60 \
--lr-scheduler single_step \
--batch-size 64 \
--fixbase-epoch 10 \
--open-layers classifier \
--transforms random_flip \
--load-weights saved-models/osnet_x1_0_imagenet.pth \
--save-dir log/osnet_x1_0-market1501-softmax \
--gpu-devices 1
```

and I loaded the imagenet pretrained model from this 
https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html

**Comment by KaiyangZhou** [comment]
it would be easier to debug if you can show the *full* log

**Comment by crossknight** [comment]
Here is the full log
[train.log-2019-07-05-16-30-37.log](https://github.com/KaiyangZhou/deep-person-reid/files/3368791/train.log-2019-07-05-16-30-37.log)

**Comment by KaiyangZhou** [comment]
which commit are you using?

**Comment by crossknight** [comment]
This
d20b63c1a07833be95b2e390513456118d3aac2c

**Comment by KaiyangZhou** [comment]
looking at your total training time and performance, I would guess your model is trained on 128x64 images and tested 256x128 images

pls try the latest commit (with version=0.8.1 in `torchreid/__init__.py`) and do come back with the new result you get

**Comment by crossknight** [comment]
ok, thank you.

I will show new results after try

**Comment by crossknight** [comment]
@KaiyangZhou 
This is the new results
```
mAP: 82.5%
CMC curve
Rank-1  : 94.2%
Rank-5  : 97.8%
Rank-10 : 98.7%
Rank-20 : 99.0%
```

commit: 1f891c1b368fe77b42d843952f5c25955fb90bbb

and a log file
[train.log-2019-07-08-22-07-27.log](https://github.com/KaiyangZhou/deep-person-reid/files/3370852/train.log-2019-07-08-22-07-27.log)

Thank you for your help.

**Comment by littleTree-96** [comment]
Hi, @KaiyangZhou 
I trained the OSNet using scripts as you released:
``` bash
python scripts/main.py \
--root $DATA \
-j 4 \
--app image \
--loss softmax \
--label-smooth \
-s $source \
-a osnet_x1_0 \
--height 256 \
--width 128 \
--optim amsgrad \
--lr 0.0015 \
--max-epoch 150 \
--stepsize 60 \
--lr-scheduler single_step \
--batch-size 64 \
--fixbase-epoch 10 \
--open-layers classifier \
--transforms random_flip random_erase \
--load-weights saved-models/osnet_x1_0_imagenet.pth \
--save-dir $save_dir \
--gpu-devices 0
```
and used `--transforms random_flip random_erase` to increase the performance. I got the results on Market1501 and Duke as you reported in the paper. On dataset Market1501, I got mAP: 84.0% and Rank-1: 94.9% (epoch 150); on dataset Duke, I got mAP: 72.1% and Rank-1: 88.1% (epoch 150). But on dataset CUHK03 (detected) I got mAP: 57.9% and Rank-1: 61.5% (epoch 150), which were much lower than you reported: mAP: 67.8% and Rank-1: 72.3%. I tried to add `--use-metric-cuhk03` during testing, but the performance was similar. Did I miss some parameters during training or testing, or I used wrong dataset? I downloaded the dataset from [https://drive.google.com/file/d/0BxJeH3p7Ln48djNVVVJtUXh6bXc/edit](url) and mat files from [https://github.com/zhunzhong07/person-re-ranking/tree/master/evaluation/data/CUHK03](url).

**Comment by KaiyangZhou** [comment]
@littleTree-96 cuhk03 is a bit tricky, use [this config](https://github.com/KaiyangZhou/deep-person-reid/blob/master/configs/im_osnet_x1_0_softmax_256x128_amsgrad_cosine.yaml)

**Comment by KaiyangZhou** [comment]
The training scripts for osnet are available at https://github.com/KaiyangZhou/deep-person-reid/tree/master/configs

**Comment by littleTree-96** [comment]
Thank you for your help. I will try this script.

**Comment by L1aoXingyu** [comment]
@KaiyangZhou Hi, Why did you only use softmax but no triplet loss for training osnet? In my opinion, joint training with triplet loss and classification loss will boost the model's performance.

**Comment by KaiyangZhou** [comment]
@L1aoXingyu 
I thought so. However, in my early experiments (w/ resnet, densenet), the improvement brought by softmax+triplet diminishes when the pretrained base network is fixed during early training (this leads to a better result in both same-dataset and cross-dataset settings).

I guess the triplet loss somehow helps correct the bad gradients produced by the randomly initialised softmax layer in the early epochs. However, this can also be achieved by the more efficient *fixbase* training.

In a more realistic case where we combine different datasets for model training, softmax gives the best result while softmax+triplet does not show any advantage.

Personally, I prefer to stick with softmax because it's powerful enough (so I can focus on the other parts). Intuitively, compared with the triplet, the softmax takes into account all training identities in a single gradient update, which is much more than what the triplet can do.

However, I do expect there might be some chance of designing some fancy losses specific for reid, e.g., the non-differentiable mAP, ...

**Comment by 9meng9** [comment]
When Osnet was trained on MSMT17 and tested on Market1501, have Osnet trained on 4101 identities or 1041 identities?

**Comment by KaiyangZhou** [comment]
@9meng9 I assume you are talking about the setting presented in https://arxiv.org/abs/1910.06827? we used 4101 ids as stated in the paper

**Comment by L1aoXingyu** [comment]
@KaiyangZhou Hi, have you tried some face loss, such as cosface or arcface? I just try arcface on osnet, and I can get a good result.

osnet_fixbase10_arcface | 88.2 (75.4)
-- | --

**Comment by KaiyangZhou** [comment]
@L1aoXingyu 
Interesting. 88.2 for which dataset/setting?

I don't remember I tried those face losses before except center loss. If the improvement is convincing, I mean you could get consistent improvements with different random seeds, then it's a good thing for reid. I always believe there is a close connection between face and reid in terms of training techniques and many face tricks can be transferred to reid. (maybe I did not tune it right to get improvement)

You are welcome to submit a PR to add the (proved effective) face losses to this repo if you want.

**Comment by L1aoXingyu** [comment]
For dukemtmc and only use arcface, fix base training for 10 epochs then joint training. I will try it on other dataset and then give a feedback.

**Comment by greatgeekgrace** [comment]
@KaiyangZhou Thanks for you code. Here is a confusion. I read the code, that all models adopt softmax+triplet loss. But In the docs (https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html) you give, the loss you only write softmax. Why. Thanks a million.

**Comment by greatgeekgrace** [comment]
@KaiyangZhou

**Comment by KaiyangZhou** [comment]
models in the model zoo are trained with softmax only

the code accepts both softmax and softmax+triplet training, depending on the self.loss

**Comment by ammad-autocanvas** [comment]
I tried training OsNet for Car Re identification. The loss is like this, Why is model not converging??
![Screenshot from 2019-11-14 19-04-52](https://user-images.githubusercontent.com/56635137/68863978-2af92c00-0712-11ea-9e96-64d23d38daf9.png)

**Comment by Leo-xxx** [comment]
[@KaiyangZhou](https://github.com/KaiyangZhou/deep-person-reid/issues/200#issuecomment-552183830) So, will the models trained with triplet loss be released in the future? 🔥

**Comment by KaiyangZhou** [comment]
@ammad-autocanvas hmm, the curve looks indeed weird, and, sexy ... are you sure this is the loss curve and not the data loading curve?

there might be some issue with your code I guess, because it's unlikely that a network cannot converge (even the data is flawed https://arxiv.org/abs/1611.03530)

**Comment by KaiyangZhou** [comment]
@Leo-xxx this is uncertain, sorry

**Comment by BarryKCL** [comment]
> @KaiyangZhou Hi, have you tried some face loss, such as cosface or arcface? I just try arcface on osnet, and I can get a good result.
> 
> osnet_fixbase10_arcface	88.2 (75.4)

can you share your details?I tried arcface on osnet_ain with dukemtmcreid,but don't get any improvement.

class CosineLoss(nn.Module):
    r"""Implement of large margin arc distance: :
    Args:
        feat_dim: size of each input sample
        num_classes: size of each output sample
        s: norm of input feature
        m: margin
        cos(theta + m)
    """
    def __init__(self, num_classes=751, feat_dim=512, s=32.0, m=0.50, easy_margin=False):
        super(CosineLoss, self).__init__()
        self.s = s
        self.m = m
        self.weight = nn.Parameter(torch.Tensor(num_classes, feat_dim))
        nn.init.xavier_uniform_(self.weight)

        self.easy_margin = easy_margin
        self.cos_m = math.cos(m)
        self.sin_m = math.sin(m)

        # make the function cos(theta+m) monotonic decreasing while theta in [0°,180°]
        self.th = math.cos(math.pi - m)
        self.mm = math.sin(math.pi - m) * m
        self.criterion = FocalLoss().to('cuda')


    def forward(self, x, label):
        # cos(theta)
        cosine = F.linear(F.normalize(x), F.normalize(self.weight.cuda()))
        # cos(theta + m)
        sine = torch.sqrt(1.0 - torch.pow(cosine, 2))
        phi = cosine * self.cos_m - sine * self.sin_m

        if self.easy_margin:
            phi = torch.where(cosine > 0, phi, cosine)
        else:
            phi = torch.where((cosine - self.th) > 0, phi, cosine - self.mm)

        #one_hot = torch.zeros(cosine.size(), device='cuda' if torch.cuda.is_available() else 'cpu')
        one_hot = torch.zeros_like(cosine)
        one_hot.scatter_(1, label.view(-1, 1), 1)
        output = (one_hot * phi) + ((1.0 - one_hot) * cosine)
        output = output * self.s
        loss = self.criterion(output, label)
        return loss

**Comment by Umraz-Hussain-MyWorld** [comment]
Hi,

Upon loading the model and configuring it with CUDA, it is expected to utilize the GPU exclusively. However, it has been observed that the model is consuming 2.3GB of GPU and an additional 4GB of CPU, which is unexpected. Further experimentation with the device parameter revealed that using the CPU resulted in lower CPU consumption (~300MB) and no GPU usage, albeit at the cost of increased processing time. The reason for the model's high usage of both CPU and GPU remains unclear. Further investigation is required to determine the root cause of this behavior. any suggestions?

below is the sample model loading code

```
from torchreid import models, utils

class ModelLoader:
    osnet_model = None  
    @staticmethod
    def get_model():
        if ModelLoader.osnet_model is None:
            ModelLoader.osnet_model = utils.FeatureExtractor(
                model_name='osnet_x1_0',
                model_path='osnet_x1_0_imagenet.pth',
                device='cuda:0'
            )
        return ModelLoader.osnet_model

# load model
model_load_time = time.time()
try:
    feature_extractor_model = ModelLoader.get_model()
    logger.debug("model is loaded")
except Exception as e:
    # print(f'{e}')
    logger.error(f"exception while loading model:   {e}") 
```
